deprecate `_remote_checksum()` and remove use in fetch

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Deprecate `_remote_checksum()` since only `fetch` was using it.
Change `fetch` to use `_exectue_remote_stat()`.
Fixes #29457.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/plugins/action/__init__.py`
`lib/ansible/plugins/action/fetch.py`

samdoran committed:
Return error message from _remote_checksum()

bcoca commented:
we could just deprecate _remote_checksum and have fetch use stat directly like all other actions

samdoran committed:
Revert "Return error message from _remote_checksum()"

This reverts commit 4b6d613d0f881a6204eb7de79e62d033b00bad11.

samdoran committed:
Remove use of _remote_checksum from fetch module

samdoran committed:
Increase test coverage

samdoran committed:
Restructure tests

samdoran committed:
Increase test coverage more and add another test run with became=True

samdoran committed:
Improve code related to fail_on_missing

samdoran committed:
Make error message consistent

samdoran committed:
Increase coverage some more

samdoran committed:
Correct logic when remote stat fails

samdoran committed:
Allow setup_remote_tmp_dir role to skip removal of tmp dir in handlers

samdoran committed:
Add tests covering the use of stat from the fetch module

This required creating an unpriveleged user account and connecting as that user remotely since
it is not possible to create a file that the root user cannot stat.

bcoca commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -803,7 +804,7 @@ def _remote_checksum(self, path, all_vars, follow=False):
         4 = stat module failed, likely due to not finding python
         5 = appropriate json module not found
-        '''
+        """
```
Add a runtime deprecation

samdoran replied:
Ok. I was wondering how we actually handle that. Do we just do a `display.deprecated()` since it's an internal call and the user might not know what to do with that information?

bcoca replied:
yep, but mostly a note that 'the plugin author should update his code'

bcoca reviewed (changes requested):
lgtm, just add runtime deprecation notice if we are deprecating the api

samdoran committed:
Test fixes

- remove /run/nologin to allow ssh logins
- make sure the known_hosts file does not get changed

samdoran committed:
Add deprecation notice to _remote_checksum()

samdoran committed:
Fix tests for macOS

macOS requires non-admin users to be added to the com.apple.access_ssh group.

There is no getent binary on macOS so use a shell command the works on Linux and macOS
to resolve the home directory.

samdoran commented:
Still working out test issues.

samdoran committed:
Use fact caching to persist remote tmp dir across playbook runs

samdoran committed:
Minor adjustment to make behavior consistent between Linux and Windows

samdoran committed:
Set known_hosts file earlier with the hope of avoiding changes to the test environment

samdoran committed:
Add comments to the test playbooks

samdoran committed:
Update deprecation message

samdoran committed:
Set ssh connection vars in inventory

Change known_hosts to write to outdir_dir instead of /dev/null. Not sure why this is failing to
take effect on Fedora 34.

samdoran committed:
Make line shorter.

samdoran committed:
Define output_dir during cleanup

samdoran commented:
/azp run
